src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1105,9 +1154,65 @@ private static ImmutableArray<StateForCase> RemoveEvaluation(ImmutableArray<Stat
             // if the tests are for unrelated things, there is no implication from one to the other
             if (!test.Input.Equals(other.Input))
-                return;
+            {
+                switch (test.Input.Source, other.Input.Source)
+                {
+                    case (BoundDagPropertyEvaluation { IsLengthOrCount: true } s1, BoundDagPropertyEvaluation { IsLengthOrCount: true } s2):
+                        {
+                            (BoundDagTemp s1Input, _, int s1Offset) = GetCanonicalInput(s1);
+                            (BoundDagTemp s2Input, _, int s2Offset) = GetCanonicalInput(s2);
+                            if (s1Input.Equals(s2Input))
+                            {
+                                Debug.Assert(whenTrueValues is not null);
+                                Debug.Assert(whenFalseValues is not null);
+                                whenTrueValues = ValueSetFactory.Shift(whenTrueValues, s1Offset - s2Offset);
+                                whenFalseValues = whenTrueValues.Complement();
+                                break;
+                            }
+                        }
+                        return;
+                    case (BoundDagIndexerEvaluation s1, BoundDagIndexerEvaluation s2):
+                        {
+                            (BoundDagTemp s1Input, BoundDagTemp s1LengthTemp, int s1Index) = GetCanonicalInput(s1);
+                            (BoundDagTemp s2Input, BoundDagTemp s2LengthTemp, int s2Index) = GetCanonicalInput(s2);
+                            if (s1Input.Equals(s2Input))
+                            {
+                                Debug.Assert(s1LengthTemp.Equals(s2LengthTemp));
+                                if (s1Index == s2Index)
+                                {
+                                    break;
+                                }
+                                if (s1Index < 0 != s2Index < 0)
+                                {
+                                    Debug.Assert(state.RemainingValues.ContainsKey(s1LengthTemp));
+                                    var lengthValues = (IValueSet<int>)state.RemainingValues[s1LengthTemp];
+                                    if (lengthValues.IsEmpty)
+                                    {
+                                        return;
+                                    }
+
+                                    int lengthValue = s1Index < 0 ? s2Index - s1Index : s1Index - s2Index;
+                                    if (lengthValues.All(BinaryOperatorKind.Equal, lengthValue))
+                                    {
+                                        break;
+                                    }
+                                    if (lengthValues.Any(BinaryOperatorKind.Equal, lengthValue))
+                                    {
+                                        trueOther = new Tests.One(new BoundDagValueTest(syntax, ConstantValue.Create(lengthValue), s1LengthTemp));
```

Comment:
I think this may be causing the problem I pointed out where a `length == 3` test is injected for a `[.., <= 0, _]` case.

Looking more at the code as part of this code review, I'm still wondering if we could leverage `Tests.OrSequence` and represent `[.., 42]` as: `((length == 1 and t2 = [0]) or (t2 = [^1]) and (t2 == 42)`. The existing machinery should be able to take it from there.
The main issue with such an approach is that we have to figure out the maximum length of list-pattern so we know how many branches to emit for the `OrSequence`: ie. `((length == 1 and t2 = [0]) or (length == 2 and t2 = [1]) or (t2 = [^1])) and ...` when we should track 3 distinct sub-cases.

Reply:
>I think this may be causing the problem I pointed out where a length == 3 test is injected for a [.., <= 0, _] case.

This test is expected to get subsumed. I removed it from the opposite branch and added an assert.

Reply:
<details><summary>new decision dag dump</summary>

```
State 0
    1. [case [_, > 0, ..]:] AND(?DagNonNullTest(t0), t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t2=DagIndexerEvaluation(t0[1]), ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32)))
    2. [case [.., <= 0, _]:] AND(?DagNonNullTest(t0), t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t3=DagIndexerEvaluation(t0[-2]), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    Test: ?DagNonNullTest(t0)
    TrueBranch: 1
    FalseBranch: 10
State 1
    1. [case [_, > 0, ..]:] AND(t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t2=DagIndexerEvaluation(t0[1]), ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32)))
    2. [case [.., <= 0, _]:] AND(t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t3=DagIndexerEvaluation(t0[-2]), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    Test: t1=DagPropertyEvaluation(t0.Length)
    TrueBranch: 2
State 2
    1. [case [_, > 0, ..]:] AND(?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t2=DagIndexerEvaluation(t0[1]), ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32)))
    2. [case [.., <= 0, _]:] AND(?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t3=DagIndexerEvaluation(t0[-2]), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    Test: ?DagRelationalTest(t1 >= ConstantValueI32(2: Int32))
    TrueBranch: 3
    FalseBranch: 10
State 3 REMAINING t1:[2..2147483647]
    1. [case [_, > 0, ..]:] AND(t2=DagIndexerEvaluation(t0[1]), ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32)))
    2. [case [.., <= 0, _]:] AND(t3=DagIndexerEvaluation(t0[-2]), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    Test: t2=DagIndexerEvaluation(t0[1])
    TrueBranch: 4
State 4 REMAINING t1:[2..2147483647]
    1. [case [_, > 0, ..]:] ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32))
    2. [case [.., <= 0, _]:] AND(t3=DagIndexerEvaluation(t0[-2]), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    Test: ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32))
    TrueBranch: 5
    FalseBranch: 6
*State 5 REMAINING t2:[1..2147483647] t1:[2..2147483647]
    1. [case [_, > 0, ..]:] MATCH
State 6 REMAINING t2:[-2147483648..0] t1:[2..2147483647]
    2. [case [.., <= 0, _]:] AND(t3=DagIndexerEvaluation(t0[-2]), OR(?DagValueTest(t1 == ConstantValueI32(3: Int32)), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32))))
    Test: t3=DagIndexerEvaluation(t0[-2])
    TrueBranch: 7
State 7 REMAINING t2:[-2147483648..0] t1:[2..2147483647]
    2. [case [.., <= 0, _]:] OR(?DagValueTest(t1 == ConstantValueI32(3: Int32)), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    Test: ?DagValueTest(t1 == ConstantValueI32(3: Int32))
    TrueBranch: 9
    FalseBranch: 8
State 8 REMAINING t2:[-2147483648..0] t1:[2..2],[4..2147483647]
    2. [case [.., <= 0, _]:] ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32))
    Test: ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32))
    TrueBranch: 9
    FalseBranch: 10
*State 9 REMAINING t2:[-2147483648..0] t1:[2..2147483647]
    2. [case [.., <= 0, _]:] MATCH
*State 10 FAIL
```

</details>